Sc/refsview

scunz opened the pull request:
An alternative approach to improve the RefsView.

This is based upon ngf/model-update and includes the compile-fix commit, which should be removed from the branches by rebasing them to development.

This approach doesn't consider stashes and namespaces yet. The support for stashes is lacking in RepoMan and probably in GitWrap. For namespaces I'm simply not sure how to display them correctly. The basic problem is, that each namespace could have both branches and tags (as well as other namespaces, in fact).

But we're actually having the same problem with remotes, where we _should_ show a separate tree for tags.

antis81 committed:
update history model, when a git reference was created

antis81 committed:
privatized appending history entries

antis81 committed:
moved assertion to where it belongs

antis81 committed:
simplified readability of HistoryModel::scanInlineRefs()

antis81 committed:
added some documentation

antis81 committed:
Ref-Model: added method to lookup a RefScope item for a Git::Reference

antis81 committed:
Ref-Model: added internal method to get a model index for a random RefItem

antis81 committed:
Ref-Model: added internal methods to insert a Git::Reference into the tree

antis81 committed:
Ref-Model: actually clear the reference tree in case of a model reset

antis81 committed:
Ref-Model: outsourced creation of references during model reset

antis81 committed:
Ref-Model: always create the RefScope items in a fixed order

antis81 committed:
Ref-Model: react to added references using RM::EventInterface

This is to be optimized. Currently it triggers a "re-lookup" of the added reference, using its full qualified name. Instead, the RM::Ref object should be used directly. This needs a bigger re-design of the current code model.

antis81 committed:
minor cleanup

antis81 committed:
RefsView: expand tree in initial state

antis81 committed:
Commit Dialog: Show errors during commit as warning message.

antis81 committed:
History-Diff: Moved "diff to parent commits" to top.

This sets "Diff To: All parents" as "default" behaviour. Technically, there is no default, but it is the first entry in the list. Selecting a default behaviour is a possible candidate for "History Settings".

antis81 committed:
RefsView: added RefsViewDelegate

antis81 committed:
RefsView: Set item delegate for tree view.

antis81 committed:
RefsView: return the base color for RowBgGradientRole

The gradient will be created in the item delegate.

antis81 committed:
RefsView: let the row background be drawn by the item delegate

antis81 committed:
RefItem: removed unused parameter

antis81 committed:
RefsView: first steps to migrate to RM::Ref

antis81 committed:
updated license info

antis81 committed:
fixup

antis81 committed:
History-Module: added "less than" comparator for "inline references"

Also added the mIsDetached flag, that will be used in upcoming commits.

antis81 committed:
HistoryModel: outsourced internal update of "inline references"

antis81 committed:
HistoryModel: fixed comparison of "inline references" while refreshing

antis81 committed:
minor cleanup

antis81 committed:
HistoryModel: minor optimizations

antis81 committed:
HistoryModel: sort "inline references"

antis81 committed:
updated license info

antis81 committed:
HistoryModel: add an "inline reference" for a detached HEAD

antis81 committed:
HistoryModel: Use Qt5 QStringLiterals -> This breaks compatibility to Qt4

We don't need Qt4-Compatibility anyways and I don't want to keep that nasty stuff around.

antis81 committed:
HistoryModel: a stash is never "active"

antis81 committed:
HistoryModel: minor optimization

antis81 committed:
HistoryModel: react to symbolic reference changes

This is currently not triggered and needs work in RM::RepoMan.

antis81 committed:
HistoryView: mark a detached HEAD in a red color

antis81 committed:
removed double comparison of mRefName

antis81 committed:
moved doxygen comments to cpp

antis81 committed:
HistoryModel: fixed temporary copy of object id

antis81 committed:
HistoryModel: changed back stash status to "isCurrent", as it contains the most recent stash

antis81 committed:
HistoryModel: react to refMoved signal

Most times this signal is send from RM when the HEAD reference changes.

antis81 committed:
HistoryModel: replaced less than comparator with operator <()

antis81 committed:
replaced QLatin1Literal with QStringLiteral

antis81 committed:
removed obsolete slot "onRefLinkChanged", which is handled by "onRefMoved"

antis81 committed:
Reference-Model: react to refMoved event

Currently forces a refresh of all the model data.

antis81 committed:
Reference-View: removed workaround to update the view manually

antis81 committed:
RepoManLogger: removed obsolete event "refLinkChanged"

antis81 committed:
History-Model: moved docs to source

antis81 committed:
History-Model: refresh inline references on a removed reference

This also is valid for a renamed reference!

antis81 committed:
Refs-Model: removed unused code

antis81 committed:
RefItem: added temporary helper methods

antis81 committed:
RefItem: removed obsolete method "setData"

antis81 committed:
Refs-Model: added temporary helper method to find invalid references

antis81 committed:
Refs-Model: react to deleted references

This is also valid for a renamed reference!

Minor downside: When a scope/namespace has no more children, it will still stay visible.

antis81 committed:
Refs-View: We don't need to manually update the RefsView anymore - finally!

antis81 committed:
minor: removed debug output

antis81 committed:
Refs-Model: no need to check "ref" and "invalidRefs"

antis81 committed:
Refs-Model: inlined inline methods

antis81 committed:
RepoMan-Events: check, if the event-repo equals the model's repo

antis81 committed:
Refs-Model: made "scopeForRef" future proof

antis81 committed:
minor: updated copyright info

scunz committed:
Remove RefsSortProxy

scunz committed:
Don't expose Model's internals

scunz committed:
Move RefItem file

scunz committed:
Create RefRoot, so we can have pure virtual methods in RefItem.

scunz committed:
Don't use data() to transport the type

scunz committed:
Sort methods more logical

scunz committed:
Remove totally unused "column" from RefItem::data()

scunz committed:
Make Headlines a bit taller

scunz committed:
Rewrite BranchesModel to make use of RepoMan

antis81 commented:
Here's what it looks like in Ubuntu :+1:
![mgv-ref-tree-sc](https://cloud.githubusercontent.com/assets/440517/6160161/2e6a077c-b25f-11e4-9793-2e050b3fea5d.png)

I just compiled and didn't look deeper into this code yet. Btw. Compared to the other branch, drawing with current code is actually performing faster, which you can easily see by dragging the splitters.

For the visual appearance, I have some questions / notes:
- I think, the icons are "cluttering" the view. Can you try putting them only to the headers?
- (Icons for `RefTreeNodes` should be kept, I think.)
- I really like to have the colors back. But, it's not finished, is it?
- Same for the bold font, that marks the active branch.

scunz commented:
> Btw. Compared to the other branch, drawing with current code is actually performing faster, which you can easily see by dragging the splitters.

Which code are you talking about? _this_ code omits your reading of the repository from disc on each drawing, so it _is_ faster.

> I think, the icons are "cluttering" the view. Can you try putting them only to the headers?
> (Icons for RefTreeNodes should be kept, I think.)

No, these icons don't make any sense in the header. And having an icon for all items finally removes that annoying cluttering that made it impossible to spot the actual hierarchy. Everything is now aligned as it usually is in a tree.

> I really like to have the colors back. But, it's not finished, is it?

No, this isn't done. The underlying data model doesn't carry that information and I don't want to reintroduce the code that reads this information from disc on every drawing operation.

> Same for the bold font, that marks the active branch.

Same cause, same solution. The model has to carry this data, determined _once_ during a refresh and updated with events accordingly.

antis81 commented:
> Which code are you talking about? this code omits your reading of the repository from disc on each drawing, so it is faster.

Yes, this is exactly what I was thinking about :). There's some other aspects to: i.e. you kicked the sort model layer. But this is the main point. It's not too bad, as the lookup is always done on the items drawn on screen (which is always <50). Ah ... I'm not writing any more, until I reviewed the code :).

> ... And having an icon for all items finally removes that annoying cluttering that made it impossible to spot the actual hierarchy. Everything is now aligned as it usually is in a tree.

I'll adapt to this and see what future brings. :smile:

> The underlying data model doesn't carry that information and I don't want to reintroduce the code that reads this information from disc on every drawing operation.

Happy to hear that. We need to discuss the `RM::Repo::heads()` stuff, I introduced in `libMacGitverCore -> PR 22/23` and make it right :). I.e.: Keep the HEAD-ref as a separate instance - out of the collections. Lookup and compare would be fast, furious and even cool :).

In current implementation that means:

``` c++
QVariant RefBranch::data(int role) const
{
    Q_ASSERT(mObject);

    switch (role) {
    case Qt::FontRole:
        if( isCurrentBranch() )
        {
            QFont f;
            f.setBold( true );
            return f;
        }
        #endif
        break;

    case RefItem::RowBgGradientRole:
        if ( equalsHeadId() )
        {
            QColor back = isCurrentBranch()
                          ? QColor::fromHsl(35, 255, 190)
                          : QColor::fromHsl(35, 255, 190).lighter(130);
            return back;
        }
        #endif
        break;

    case Qt::EditRole:
        return object()->name();
    }

    return RefItemObject::data(role);
}

// spice it up with some private salt & pepper :)

const RM::Ref* RefBranch::findHEAD( const RM::Repo* repo ) {
    // this can be optimized in libMacGitverCore, by providing a RM::Repo::head() method
    // Anyways, this code will work!

    RM::Ref* HEAD = NULL;
    const QString headStr = QStringLiteral("HEAD");
    foreach ( RM::Ref* candidate, repo->heads().childObjects<RM::Ref>() ) {
        if ( candidate.fullName() == headStr ) {
            HEAD = candidate;
            break;
        }
    }

    return HEAD;
}

bool RefBranch::isCurrentBranch() const
{
    const RM::Repo* repo = mObject->repository();
    const RM::Ref* HEAD = findHEAD( repo );
    // Q_ASSERT( HEAD );

    return HEAD && HEAD->symbolicTarget() == object()->fullName();
}

bool RefBranch::equalsHeadId() const
{
    const RM::Repo* repo = mObject->repository();
    const RM::Ref* HEAD = findHEAD( repo );
    // Q_ASSERT( HEAD );

    return HEAD && HEAD->id() == object()->id();
}
```

antis81 commented on RefsViews/Branches/RefItem.cpp:
```diff
@@ -0,0 +1,199 @@
+}
+
+#if 0
+RefNameSpace::RefNameSpace(RefItem *p, const QString &t)
```
We can safely remove this class!

antis81 commented on RefsViews/Branches/RefItem.cpp:
```diff
@@ -0,0 +1,199 @@
+        break;
+
+    case Qt::EditRole:
+        return object()->name();
```
This is wrong! We need the reference shorthand here. How to get that without reading from disk? Simple:

``` c++
const Git::RefName rn( object()->fullName() );
return rn.shorthand();
```

antis81 replied:
In the old code, the `Qt::EditRole` code was unused, as the name was set directly in the rename dialog. Look [here](https://github.com/macgitver/MacGitverModules/blob/development/RefsViews/RefRenameDialog.cpp#L72) for comparison.

scunz committed:
Remove EditRole

scunz committed:
Fix regresion in RenameDlg

scunz committed:
Make RenameDialog more intuitive

scunz committed:
Refresh Repositories

scunz committed:
Ref Rename Dialog doesn't need to depend on Model internals

scunz committed:
Now, remove the manual indentation to get the tree aligned exactly

scunz committed:
Add missing spacer to Rename Ref dialog

scunz committed:
Keep Headlines disabled and not selectable

As far as _behaviour_ is concerned, we don't want the headlines enabled.
So that they cannot actually even be focused. However, this is certainly
not what we want to draw.

antis81 commented:
This needs rebasing due to the macro changes in `GitWrap`. Otherwise, compiler will complain about a missing semicolon.

scunz commented:
> This needs rebasing due to the macro changes in GitWrap. Otherwise, compiler will complain about a missing semicolon.

Yepp. Though my local branch still has that commit around - I never start to develop on the "development" branch; not even for fixes :)

scunz commented:
Rebase done :)

antis81 commented:
> ... I never start to develop on the "development" branch; not even for fixes :)

Maybe that's why lg2 renamed their "development" branch back to "master" :smile:.

> Rebase done :)

It's funny to see the website updating while writing in this box :smile:. Thanks!

scunz committed:
Show current branch in bold
